### ISSUE TITLE
would like git status supported in Rowan

### ISSUE THREAD

**Issue by ericwinger** [GemTalk/Jadeite#323]
When I'm working in Rowan code, I try to commit locally often so as not to lose my work. (a trick I learned from @dalehenrich). I don't want to push to git until I've tested those changes. 
However, sometimes it takes awhile to test those changes and I forget to push to github. Jadeite doesn't currently show me any indication that my commits are ahead of github. 

It might be useful to support ``git status`` in Rowan which could be shown in Jadeite. I don't see that it's there today - at least there's no Rowan method selector ``status`` or ``gitStatus``. 

I'm open to other ways to do this.

**Comment by dalehenrich** [comment]
this seems to be a jadeite issue that might end up with needed Rowan support

**Comment by LisaAlmarode** [comment]
You can do it now using an expression such as
```
Rowan gitTools 
	gitstatusIn: ((Rowan image loadedProjectNamed: 'VegTest') asDefinition repositoryRootPath) with: ''
```
Eric, if this isn't what you want please clarify.

This gets git status, but deciphering it is not all that easy. I can imagine a nice API but it would be a fair bit of work.

**Comment by dalehenrich** [comment]
for tODE I use the following variant of `git status`
```smalltalk
gitStatusFor: gitPath
  "answer emtpy string if repo has not been modified"

  | command result |
  command := 'cd ' , gitPath , '; git status --porcelain'.
  result := System performOnServer: command.
  (result beginsWith: 'fatal:')
    ifTrue: [ ^ nil ].
  ^ result
```
and it is called by `isGtDirty`:
```smalltalk
isGitDirty
  self hasGitRepository
    ifFalse: [ ^ false ].
  ^ (MetacelloPlatform current gitStatusFor: self gitRepoDirectoryPath)
    ifNil: [ ^ false ]
    ifNotNil: [ :response | ^ response isEmpty not ]
```
which is a pretty simple way to determine whether or not the underlying git repo has uncommitted writes.

Here's a screenshot of the tODE project list:
![image](https://user-images.githubusercontent.com/623951/49170500-206f3800-f2f1-11e8-900e-63d4df0b0c6a.png)
The `+` means that the git repo is dirty, the `*` means the project is dirty...

**Comment by dalehenrich** [comment]
It looks the following could be written without needing any additional tool support:
```smalltalk
isGitDirty: repositoryRootPath
  "use Lisa's code for repositoryRootPath ... plus additional logic for non-git repos ... "
  [ ^ (Rowan gitTools 
      gitstatusIn: repositoryRootPath 
      with: '--porcelain') isEmpty not ] on: Error do: [:ignored | ^ false ]
```

**Comment by ericwinger** [comment]
Will need to add Jadeite support. Reopening and assigning to myself.

**Comment by ericwinger** [comment]
@dalehenrich I tried the code [here ](https://github.com/GemTalk/Jadeite/issues/323#issuecomment-442545934) but it didn't seem to tell me what I hoped. I hoped it would tell me if my local repository had a commit that was not pushed to the github repository.  Instead, it shows me if I have uncommitted changes written to the local repository. 

While knowing if the local disk is dirty is helpful and I can incorporate it into Jadeite, I would still like to know if a local commit has been pushed to remote. Is it possible to know, using Rowan tools today, if a local commit has been pushed? 

Lisa's code [here](https://github.com/GemTalk/Jadeite/issues/323#issuecomment-442267940) tells me the status in English, but it's not possible to build a reliable indicator from the text.

**Comment by dalehenrich** [comment]
It looks like this just might do the trick (returns empty status if the commits have been pushed):
```
git log <remote>/<branch-name>..<branch-name>
```
Of course, this is all relative to the remote ... and a repository can have multiple remotes, so not sure if it is **feasible** to do what you want ...

**Comment by ericwinger** [comment]
> not sure if it is feasible to do what you want

That's a good point. Have to think a bit on that. Might be able to show something, even if it's limited. 
In the meantime, added support for showing if the project is dirty on local disk.

**Comment by dalehenrich** [comment]
In v2, I have a field for the remote ... IMHO, knowing whether or not  a project is dirty on disk is right up there with project skew .... you don't want to overwrite changes on disk unless you are keeping track ... ... if you're not sure whether or not you have pushed, you can always do a push *just in case* and it's pretty harmless ... personally I puah right after a commit, if I'm going to push...

**Comment by ericwinger** [comment]
Now show local disk dirty status with an asterisk. Removed notion of default project and re-purposed project asterisk. Fixed in 3.1.0. 

As discussed in thread, no longer seeking status of "have I pushed this commit to git".

**Comment by LisaAlmarode** [comment]
I've been seeing the asterisks next to the Rowan project immediately after pull and load.  This is bothering me since I am, and should be able to confirm, that I don't have a change from the origin.

On executing the git status command, I think I can see why:

Rowan gitTools 
	gitstatusIn: ((Rowan image loadedProjectNamed: 'Rowan') asDefinition repositoryRootPath) with: '' 'HEAD detached at Oscar-3.1.0pre-release2
Untracked files:
  (use "git add <file>..." to include in what will be committed)

	cypress.out
	hints
	rowan.out
	ston.out
	tonel.out
	updateRowan

nothing added to commit but untracked files present (use "git add" to track)
'

When I delete/remove these files the asterisk goes away.

I don't like this behavior.  Technically this is probably correct... although I think the presence of untracked files could be ignored.  But the problem is that there's no way to figure out why it's got the asterisk, if you don't know the query above, it's just a stubborn asterisk.  

Perhaps we need to put a menu item somewhere, for the query to get the text details of the status, so while you can see at a glance what's going on you do have access to details.

**Comment by dalehenrich** [comment]
@LisaAlmarode, I disagree that untracked files should be ignored ... In tODE, I will make changes to disk files (like adding a new class to a package) and then go back into Rowan and do additional work that I save to disk  and knowing that there are untracked files on disk is a caution, that when I save my packages, I could destroy my manually created files ... also when you do a commit from Jadeite, I assume that the code currently does a `git add` since normal writes can create untracked files ... which means that those *.out files lying around could be committed whether you want them to be committed or not ... 

Sooo, it might be annoying to have meaningless (from your perspective) `+` popping up, but for me I want to be warned that there are modified/untracked files on disk, and that I'd better understand what's down there before doing anything else ...

**Comment by dalehenrich** [comment]
It is interesting to note that `gitk` doesn't notice that there are untracked files on disk, but then `gitk` does not do `git commit` or `git add`, either...

**Comment by ericwinger** [comment]
The asterisk comes from Rowan and means that the it thinks the disk is dirty. (The indicator `*` was a re-purposing of the unused default project notion which was retired.) 

So if you see an asterisk, it comes from this code: 

```
RowanProjectService>>isGitDirty
  "From https://github.com/GemTalk/Jadeite/issues/323#issuecomment-442545934"

  [ 
  ^ (Rowan gitTools gitstatusIn: self repositoryRootPath with: '--porcelain')
    isEmpty not ]
    on: Error
    do: [ :ignored | ^ false ]
```
@dalehenrich may have some ideas on why `pull` and `load` show as dirty for you. 
`<guess>`Perhaps because you're checking out on a tag?`</guess>`

**Comment by LisaAlmarode** [comment]
Like I said, I know why the asterisk is there, and don't really object to the untracked files causing an asterisk.  

If it's there, though, I need a way to find out that it's untracked files (not a problem for the end user for the case of Rowan), and not an actual difference from the origin (a serious problem for the case of Rowan).  

The project list in general for the average end-user will include projects you are working on in which you need one level of status details, e.g. untracked files, but it will also include projects you are not working on and you need somewhat different kinds of information.  Dale and Eric don't face the second case themselves... but it needs to be considered.

**Comment by ericwinger** [comment]
Essentially, you'd like `git status` exposed in Jadeite @LisaAlmarode ?

**Comment by ericwinger** [comment]
I'm not seeing the problem after I did some more work in this area for #669 in pre-release3.  Please review again when that is released.

**Comment by ericwinger** [comment]
Closed by mistake.
